### ISSUE TITLE
Request for help for LSHSelfAttention()

### ISSUE THREAD

**Issue by andreabac3** [lucidrains/reformer-pytorch#123]
Hi @lucidrains thank you for your excellent work (I star it).

I am trying to use the LSHSelfAttention() layer in my network instead of my transformer encoder layer.

A pseudocode of what I am doing is that:
```python3
word_embeddings = word_embeddings(input)  # batch, seq_len, emb_dim
lsh_encoded = self.lsh_self_attention(word_embeddings)
```
I continuously get a vector of NaN values, to avoid it I decrease my learning rate from 1e-3 to 1e-5, but nothing is changed.

1) Am I using the correct layer?
2) Should I use Reformer() instead of LSHSelfAttention()?
I tried to use Reformer() but I also get an error there, which tells me that my sequence is divisible by the number of buckets (I'm still working on it).

**Comment by lucidrains** [comment]
It's hard to say, the NaN could result from any one of the operations in the network. 3e-4 or below for learning rate should be fine

**Comment by andreabac3** [comment]
Hi @lucidrains, thanks for your reply.

I checked and the vector of NaN is the result of applying the LSH self attention layer. I can get just a couple of batches correctly encoded and then get the NaNs and this behavior happens with all 3 learning rates I tested (I will test with 3e-4).

Regarding questions 1 and 2, am I using LSHSelfAttention() correctly or should I be using Reformer()?

Kind regards,
Andrea

**Comment by lucidrains** [comment]
@andreabac3 could you show me the context in which you placed the LSHSelfAttention code? maybe a snippet +-50 lines

**Comment by andreabac3** [comment]
Hi @lucidrains ,
thanks again for your support.

I have attached the declaration of the class and the use of the layer in the forward function which is used at the beginning of the function.

My setup is pytorch 1.5.1, pytorch lightnining 0.8.5, torch.version.cuda '10 .2 'and nvidia apex for 16bit tensors and my OS is manjaro linux.

As you can see I have a boolean called self.use_reformer with which I enable and disable the use of the reformer instead of the transformer. With the transformer everything is fine, my model has already been tested and used for other research, my aim now is to try to improve the f1 score through this new layer.

For debugging I added a control on the NaN and as you can see I get them immediately after using the reformer. I tested this with all the learning rates I mentioned in previous posts.
In this link, I have attached a print of the tensor resulting from the layer reformer https://pastebin.com/YxesSk96

```python3
import pytorch_lightning as pl
import torch
import torch.nn as nn
from reformer_pytorch import LSHSelfAttention, Reformer
from layers.positional_encoding import PositionalEncoding

class SRLModel(pl.LightningModule):

    def __init__(self, hparams, padding_token_id=0, padding_target_id=-1):
        super(SRLModel, self).__init__()
        self.word_encoder = WordEncoder(self.hparams)  # a bert wrapper 
        word_embedding_size: int = self.word_encoder.word_embedding_size  # 512
        self.reformerAttention = LSHSelfAttention(
            dim=word_embedding_size,
            )
        self.use_reformer: bool = hparams.use_reformer
        self.positional_encoding = PositionalEncoding(word_embedding_size)  # for transformer layer
        self.transformer_encoder = torch.nn.TransformerEncoderLayer(word_embedding_size, nhead=8)
       ...
```
```python3
    def forward(self, x):
        word_ids = x['word_ids']
        subword_indices = x['subword_indices']
        sequence_lengths = x['sequence_lengths']
        tokenized_sequence_lengths = x['tokenized_sequence_lengths']
        predicate_indices = x['predicate_indices']

        word_embeddings = self.word_encoder(word_ids, subword_indices=subword_indices, 
        sequence_lengths=tokenized_sequence_lengths)  # shape (batch, seq_len, hidden dim) = torch.Size([32, 63, 512])
        print(torch.isnan(word_embeddings).any())  # tensor(False, device='cuda:0')
        if self.use_reformer:
            sequence_states = self.reformerAttention(word_embeddings)  # shape torch.Size([32, 63, 512])
            print(torch.isnan(sequence_states).any())  # tensor(True, device='cuda:0')
        else:
            word_embeddings = word_embeddings.permute(1, 0, 2)  # from torch.Size([32, 63, 512]) to torch.Size([63, 32, 512])
            word_embeddings = self.positional_encoding(word_embeddings)
            sequence_states = self.transformer_encoder(word_embeddings)  
            sequence_states = sequence_states.permute(1, 0, 2) 
            print(torch.isnan(sequence_states).any())  # tensor(False, device='cuda:0')
```

Kind regards,
Andrea

**Comment by lucidrains** [comment]
@andreabac3 Hi Andrea, I spot your error!

So although transformers are just layers of attention, they still need two ingredients to work. First, everything going into the branches (the attention or feedforwards) need to be normalized. The most common normalization used is Layer normalization, and can be used as so

```python
self.norm = nn.Layernorm(hparams.model_dim)
```

```python
sequence_states = self.reformerAttnetion(self.norm(word_embeddings))
```

Also, make sure that the `sequence_states` that come out is added to the residual.

```python
x = word_embeddings
attn_out = self.reformerAttnetion(self.norm(x))
x = attn_out + x
```

**Comment by lucidrains** [comment]
@andreabac3 also, I believe to compare it to the TransformerEncoder, you'll need to use the `Reformer` class, with `causal` set to `False`

**Comment by andreabac3** [comment]
@lucidrains Hi Phil,
thanks for the clear explanation,
I added Layernorm declaration in the class constructor e tested in the forward 
```python3
x = word_embeddings
print(torch.isnan(x).any())  # tensor(False, device='cuda:0')
attn_out = self.reformerAttention(self.norm(x))  
print(torch.isnan(attn_out).any()) # tensor(True, device='cuda:0')
x = x + attn_out
print(torch.isnan(x).any()) # tensor(True, device='cuda:0')
sequence_states = x
```
but I am still getting nan even though I apply normalization (using different learning rate lower than 1e-3).
If you want I can try to save the tensor of my word embeddings on a pickle and send it to you.

Does the reformer class accept batches with different seq len or does it need fixed padding? I had trouble with that class as it tells me:
_AssertionError: Sequence length (63) needs to be divisible by target bucket size x 2 - 128_
I'm still working on it (I'm new to this world of attention layers, I think it's clear from the mistake you pointed out to me)

Kind regards,
Andrea Bacciu

**Comment by lucidrains** [comment]
@andreabac3 try a lower learning rate (3e-4 or lower)

also, for such a short sequence length, you don't need LSH attention. LSH attention is only good for sequences in the thousands to tens of thousands

**Comment by lucidrains** [comment]
@andreabac3 https://github.com/lucidrains/reformer-pytorch#additional-helpers the autopadder class will be helpful if you are using the `Reformer` class, but I think you may have to do the padding yourself if you are working with the `LSHSelfAttention` class

**Comment by lucidrains** [comment]
@andreabac3 oh actually Autopadder works with LSHSelfAttention as well, lol, I'm forgetting what I have built

**Comment by andreabac3** [comment]
@lucidrains I have tested with 3e-4, 3e-5, 3e-6 and nothing is change.
I also tested with the Reformer layer and the cited learning rate with Autopadder (autopadder remove the error) but I am still getting the NaN values in the attn_out vector (the nan is contained directly in attn_out before calculating the loss) 
The dataset used is CoNLL09.

**Comment by lucidrains** [comment]
@andreabac3 Hmm, interesting, it seems my past self had decided to fall back to full attention if the sequence length (63) is less than the default bucket size (64). It's interesting that you need the Autopadder at all. Are you on an older version?

How long into training until you hit NaNs? I just did a simple forward and backward pass locally and the layer is working as intended

**Comment by lucidrains** [comment]
I may need to see your entire script to be able to further help you. A colab notebook would help too, if you can set one up

**Comment by andreabac3** [comment]
@lucidrains  I am on the reformer_pytorch==1.2.3 version.
I cannot reach the backward pass because I get the NaN immediately after this
```python3
attn_out = self.reformerAttention(self.norm(x))  
print(torch.isnan(attn_out).any()) # tensor(True, device='cuda:0')
```
I will update you about the Colab, maybe I can be easily share to you my repository because the project is huge, but I have to contact my research group because I think the dataset is copyrighted. 

If you want I can pass you the tensor that I use as input for the Reformer layer.


It seems like that not for all random vectors the layer is capable of producing a result other than nan. What I said may sound strange but actually I am passing a more than legitimate vector of the size torch.Size([32, 63, 512]), I also tried with a torch.Size([32, 64, 512]) but it is the same.


Thank you @lucidrains so much for your helpfulness and kindness.

**Comment by andreabac3** [comment]
I solve the problem, I have disabled the 16-bit precision in pytorch lightning and now it's seems works.

So the repository don't support the 16-bit precision (nvidia apex).

Thank you we have solved the arcane ahahah

**Comment by lucidrains** [comment]
@andreabac3 ohh that's really weird, I have tested this with nvidia apex on many projects

do you have it set on O1 or O2?

**Comment by andreabac3** [comment]
@lucidrains I am using pytorch_lightning so I suppose looking in the default parameter in Trainer() class it's set
```python3
amp_level: str = 'O2',   # backward compatible, todo: remove in v1.0.0
```
which is located in _$HOME/.local/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py_

**Comment by lucidrains** [comment]
oo ok, give `amp_level='O1'` a try

**Comment by andreabac3** [comment]
@lucidrains I can confirm 16-bit precision of nvidia apex with `amp_level=O1` works!

Just for completeness I am attaching my current configuration

Class Constructor:
```python3
from reformer_pytorch import LSHSelfAttention, Reformer, Autopadder
self.reformerAttention = Autopadder(Reformer(
       dim=hparams.model_dim,
        depth=8,
        max_seq_len=200))
self.norm = nn.LayerNorm(hparams.model_dim)
```
Forward:
```python3
x = word_embeddings
attn_out = self.reformerAttention(self.norm(x))
x = x + attn_out
sequence_states = x
```

Thank you @lucidrains for the incredible support, I appreciated it!

**Comment by lucidrains** [comment]
Woohoo! Congrats 💯

**Comment by andreabac3** [comment]
Do you consider to add this information (about amp level) in the readme?

**Comment by lucidrains** [comment]
https://github.com/lucidrains/reformer-pytorch/commit/67dd588468959eab7c2510436a634777ce3ada18 done!
